lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,248 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package expiry
+
+import (
+	"context"
+	"log/slog"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/auth/authclient"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils/interval"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+var (
+	// scanInterval is the interval at which the expiry checker scans for access requests
+	scanInterval = time.Minute * 5
+	pendingRequestGracePeriod = time.Second * 40
+)
+
+const (
+	semaphoreName       = "expiry"
+	semaphoreExpiration = time.Minute
+	semaphoreJitter     = time.Minute
+
+	// minPageDelay is the minimum delay between processing each page of access requests
+	minPageDelay           = time.Millisecond * 200
+	accessRequestPageLimit = 100
+	maxExpiresPerCycle     = 120
+)
+
+// Config provides configuration for the expiry server.
+type Config struct {
+	// Log is the logger.
+	Log *slog.Logger
+	// Emitter is an events emitter, used to submit discrete events.
+	Emitter apievents.Emitter
+	// AccessPoint is a expiry access point.
+	AccessPoint authclient.ExpiryAccessPoint
+	// Clock is the clock used to calculate expiry.
+	Clock clockwork.Clock
+	// HostID is a unique ID of this host.
+	HostID string
+}
+
+// CheckAndSetDefaults checks required fields and sets default values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Emitter == nil {
+		return trace.BadParameter("no Emitter configured for expiry")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("no AccessPoint configured for expiry")
+	}
+	if c.Clock == nil {
+		return trace.BadParameter("no Clock configured for expiry")
+	}
+	return nil
+}
+
+// Service is a expiry service.
+type Service struct {
+	*Config
+
+	ctx      context.Context
+	cancelfn context.CancelFunc
+}
+
+// New initializes a expiry service
+func New(ctx context.Context, cfg *Config) (*Service, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Service{
+		Config: cfg,
+		ctx:    ctx,
+	}
+	return s, nil
+}
+
+// Start starts the expiry service.
+func (s *Service) Start() error {
```

Comment:
nit: `Start` functions conventionally return immediately.  This function blocks until the service exits.  Maybe call it `Run` or something like that?